### DISCUSSION TITLE
Detect explicit INPC implementation

### DISCUSSION THREAD

**Discussion by rsking** [canton7/PropertyChanged.SourceGenerator#38] [Ideas] [answered]
When explicitly implementing INPC, the fact that the event is already declared is not picked up, meaning that another event is created in the generated code.

I get that trying to find the event handler to create the method to call the event could be problematic (i.e. you can go through all the fields to find one that is `PropertyChangedEventHandler`, but what if there are multiple, etc).

I think it would be better to check to see if the current class (or any base class) implement INPC, and if they do, do not add the event.

To create the method, you could do the above search, or have an exception, that when explicitly implementing INPC, then an appropriate method to call it must be detectable.

**Comment by rsking** [comment]
``` csharp
public partial class SomeOtherViewModel : INotifyPropertyChanged
{
    private PropertyChangedEventHandler _handler;

    [Notify] private int _foo;

    event PropertyChangedEventHandler INotifyPropertyChanged.PropertyChanged
    {
        add => this._handler += value;
        remove => this._handler -= value;
    }
}
```

**Reply by canton7** [reply]
Do note that this code isn't thread-safe to two threads subscribing/unsubscribing at the same time

**Comment by canton7** [comment, marked as answer]
People do the darndest things... Should be fairly easy to add support for.

> I think it would be better to check to see if the current class (or any base class) implement INPC, and if they do, do not add the event.

I support someone adding INPC to the class but leaving the SG to define the event.

**Reply by canton7** [reply]
01218ee

**Reply by rsking** [reply]
Ah yes, I see. Someone can add INPC to the partial class, and then let the SG generate the actual event. Makes sense.

The implementation looks nice.

Thanks for the work. I was expecting to have to pick this one up myself ;).

**Comment by canton7** [comment]
@rsking Any other things you've spotted, before I cut a release? :)

**Reply by rsking** [reply]
That's the last blocker that I had. I've only got enhancements from now on ;)

**Reply by canton7** [reply]
Cool, v1.0.8 is out
